Unity-Technologies/com.unity.multiplayer.samples.coop
# Lobby service integration

### Description (*)


This PR introduces the Unity Lobby service into Boss Room in the form of the following features:
 - Create IP-based, Relay-based, public or private lobbies
 - Join either kind of Lobby, alongside with Quickjoin (in our case it would just join some lobby that is public and not locked) and ability to join with Lobby code.
 
From a technical standpoint this PR builds upon the DIScope, UpdateRunner and MessageChannel PRs and uses them.

Several important changes to the project have been made:
 - There is no option to use Photon Relay any more - this PR removes the Photon integration.
 - `ApplicationController` - a monobehaviour that lives throughout the application lifecycle and is resposnsible for it's lifecycle. We bind and dispose the core DIScope (it lives while the app is alive) in this class. Think of it as "Main" for the entire app.
 - Every scene in our code-base had a `...State` component corresponding to it - now these `State` classes actually contain bootstrap logic for the scene - they are the entrypoints into the scene, where we can bind the dependencies that will live while the scene is alive, and then will get disposed with when the scene unloads.
 - We log in anonymously into unity Authentication as soon as we load into main menu, which simplifies the latter flow by not requiring us to doubt the fact that we're logged in
 - MainMenu scene got a complete face-swap and now relies on a LobbyPanel prefab. LobbyPanel has a `LobbyUIMediator` component on it - it's the centerpoint of all the lobby-related UI actions and both create and join UI's are elements of it. 
 
The lobby integration bit consists of:

 - `LobbyServiceFacade` - the sole facade we have to interact with for lobby api calls of various kinds
 - The classes below are an implementation detail of LobbyServiceFacade
    - `LobbyAPIInterface` - a thin wrapper on top of Unity Lobby service calls that adds error handling and other neccessities
    - `JoinedLobbyContentHeartbeat` - keeps updated on the changes to the joined lobby
    - `LocalLobbyUser` - data for a local player instance (it's displayed name, ID etc)
    - `LocalLobby` - a local wrapper around a lobby's remote data.

Important:
 - this PR adds a way to specify command-line arguments for the built game. Currently we only can do:
 "-AuthProfile YOUR_VALUE", which would make the auth system use a different profile. Useful when testing with several builds on the same machine. 
 
### Related Pull Requests

https://github.com/Unity-Technologies/com.unity.multiplayer.samples.coop/pull/463
https://github.com/Unity-Technologies/com.unity.multiplayer.samples.coop/pull/462
https://github.com/Unity-Technologies/com.unity.multiplayer.samples.coop/pull/461

### Issue Number(s) (*)

MTT-2067
MTT-2068
MTT-2069
MTT-2070

Fixes issue(s):
### Manual testing scenarios

1. Pull the project
2. Open one or more extra project clones
3. Launch the game on all instances
4. Create an IP or Relay-based lobby and host it - this client will host the lobby and proceed into our Character Selection scene
5. This lobby (if public) would be visible on other peers as soon as the host successfully starts a relay allocation (otherwise the lobby, while relay allocation is still pending, is "locked" - it won't appear in any queries for lobbies).
6. The join code (however you host the game) will be displayed in the Character Selection scene and can be used to join the lobby manually (or to join a private lobby).

### Questions or comments

### Contribution checklist
